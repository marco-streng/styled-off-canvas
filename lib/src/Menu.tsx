import React from "react";
import styled from "styled-components";
import { useOffCanvas } from "./Context";

const Container = styled.div<{
  $isOpen: boolean;
  $background?: string;
  $duration?: string;
  $position?: "left" | "right";
  $width?: string;
}>(
  ({
    $isOpen,
    $background = "#fff",
    $duration = "500ms",
    $position = "right",
    $width = "300px",
  }) => `
  background-color: ${$background};
  height: 100%;
  left: ${$position === "left" ? ($isOpen ? 0 : "-100%") : "inherit"};
  max-width: ${$width};
  overflow-y: auto;
  position: fixed;
  right: ${$position === "right" ? ($isOpen ? 0 : "-100%") : "inherit"};
  top: 0;
  transition: ${$position} ${$duration} ease-in-out;
  width: 100%;
  z-index: 10;
`
);

const isCSR = typeof window !== "undefined";

const initalBodyStyle = {
  maxHeight: isCSR ? window.document.body.style.maxHeight : "none",
  overflow: isCSR ? window.document.body.style.overflow : "visbile",
};

/**
 * Menu component
 *
 * @param   {string}    className styled-components className for custom styling
 * @returns {component}           React component
 */
export const Menu = ({
  children,
  background,
  duration,
  position,
  width,
  closeOnEsc,
  className,
  ...props
}: {
  children: React.ReactNode;
  background?: string;
  duration?: string;
  position?: "left" | "right";
  width?: string;
  closeOnEsc?: boolean;
  className?: string;
}) => {
  const { isOpen, close } = useOffCanvas();

  // Avoid scrolling on content when the navigation is open
  React.useEffect(() => {
    const bodyElStyle = window.document.body.style;

    if (isOpen) {
      bodyElStyle.maxHeight = "100%";
      bodyElStyle.overflow = "hidden";
    } else {
      bodyElStyle.maxHeight = initalBodyStyle.maxHeight;
      bodyElStyle.overflow = initalBodyStyle.overflow;
    }
  }, [isOpen]);

  if (closeOnEsc) {
    React.useEffect(() => {
      const handleKeyDown = (event: KeyboardEvent) => {
        if (event.code.toLowerCase() === "escape") {
          close();
        }
      };

      isOpen
        ? window.document.addEventListener("keydown", handleKeyDown)
        : window.document.removeEventListener("keydown", handleKeyDown);
    }, [isOpen]);
  }

  return (
    <Container
      className={className}
      $background={background}
      $duration={duration}
      $isOpen={isOpen}
      $position={position}
      $width={width}
      {...props}
    >
      {children}
    </Container>
  );
};
