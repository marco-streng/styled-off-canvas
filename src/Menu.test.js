import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import Provider from './Provider'
import Menu from './Menu'

describe('Menu', () => {
  test('default', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen><Menu><div>content</div></Menu></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('closed', () => {
    const { container } = render(<Provider onClose={() => {}}><Menu><div>content</div></Menu></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('custom color', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen menuBackground='#000'><Menu><div>content</div></Menu></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('custom duration', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen duration='1s'><Menu><div>content</div></Menu></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('position left', () => {
    const { container } = render(<Provider onClose={() => {}} isOpen position='left'><Menu><div>content</div></Menu></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('position left closed', () => {
    const { container } = render(<Provider onClose={() => {}} position='left'><Menu><div>content</div></Menu></Provider>)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('closeOnEsc callback', () => {
    const cb = jest.fn()
    const { container } = render(<Provider isOpen onClose={cb}><Menu><div>content</div></Menu></Provider>)

    fireEvent.keyDown(container, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27
    })

    expect(cb).toHaveBeenCalled()
  })

  test('closeOnEsc callback when closed', () => {
    const cb = jest.fn()
    const { container } = render(<Provider onClose={cb}><Menu><div>content</div></Menu></Provider>)

    fireEvent.keyDown(container, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27
    })

    expect(cb).not.toHaveBeenCalled()
  })

  test('closeOnEsc callback with another key', () => {
    const cb = jest.fn()
    const { container } = render(<Provider isOpen onClose={cb}><Menu><div>content</div></Menu></Provider>)

    fireEvent.keyDown(container, {
      key: 'a',
      code: 'KeyA',
      keyCode: 65,
      charCode: 65
    })

    expect(cb).not.toHaveBeenCalled()
  })

  test('without closeOnEsc', () => {
    const cb = jest.fn()
    const { container } = render(<Provider closeOnEsc={false} isOpen onClose={cb}><Menu><div>content</div></Menu></Provider>)

    fireEvent.keyDown(container, {
      key: 'Escape',
      code: 'Escape',
      keyCode: 27,
      charCode: 27
    })

    expect(cb).not.toHaveBeenCalled()
  })

  test('body styling', () => {
    const { rerender } = render(<Provider onClose={() => {}} isOpen><Menu><div>content</div></Menu></Provider>)

    expect(document.body.style._values).toEqual({
      'max-height': '100%',
      overflow: 'hidden'
    })

    rerender(<Provider onClose={() => {}}><Menu><div>content</div></Menu></Provider>)
    expect(document.body.style._values).toEqual({})
  })
})
