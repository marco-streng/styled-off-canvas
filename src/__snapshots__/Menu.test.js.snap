// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu closed 1`] = `
<div
  className="Menu__Container-sc-mw2r80-0 kVIhPL"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu custom color 1`] = `
<div
  className="Menu__Container-sc-mw2r80-0 kyzbJQ"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu custom duration 1`] = `
<div
  className="Menu__Container-sc-mw2r80-0 iUsrZu"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu default 1`] = `
<div
  className="Menu__Container-sc-mw2r80-0 iUsrZu"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu position left 1`] = `
<div
  className="Menu__Container-sc-mw2r80-0 hQHVnN"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;

exports[`Menu position left closed 1`] = `
<div
  className="Menu__Container-sc-mw2r80-0 BpWuI"
  width="300px"
>
  <div>
    content
  </div>
</div>
`;
